_posts/2020-01-24-numpy-broadcasting/numpy-broadcasting.Rmd

```diff
@@ -0,0 +1,640 @@
+
+```{python}
+x = np.array([[1, 2], [3, 4], [5, 6]])
+x
+# array([[1, 2],
+#        [3, 4],
+#        [5, 6]])
+
+x[1, :] 
+# array([3, 4])
+```
+
+While this, arguably, makes for the easiest way to achieve that result and thus, would be the way you'd write it yourself, it's good to know that these are two alternative ways that do the same:
+
+```{python}
+x[1] 
+# array([3, 4])
+
+x[1, ] 
+# array([3, 4])
+```
+
+While the second one sure looks a bit like R, the mechanism is different. Technically, these `start:stop` things are parts of a Python _tuple_ -- that list-like, but immutable data structure that can be written with or without parentheses, e.g., `1,2` or `(1,2`) --, and whenever we have more dimensions in the array than elements in the tuple NumPy will assume we meant `:` for that dimension: Just select everything.
+
+We can see that moving on to three dimensions. Here is a 2 x 3 x 1-dimensional array:
+
+```{python}
+x = np.array([[[1],[2],[3]], [[4],[5],[6]]])
+x
+# array([[[1],
+#         [2],
+#         [3]],
+# 
+#        [[4],
+#         [5],
+#         [6]]])
+
+x.shape
+# (2, 3, 1)
+
+```
+
+In R, this would throw an error, while in Python it works:
+
+```{python}
+x[0,]
+#array([[1],
+#       [2],
+#       [3]])
+```
+
+In such a case, for enhanced readability we could instead use the so-called `Ellipsis`, explicitly asking Python to "use up all dimensions required to make this work":
+
+```{python}
+x[0, ...]
+#array([[1],
+#       [2],
+#       [3]])
+```
+
+We stop here with our selection of essential (yet confusing, possibly, to infrequent Python users) Numpy indexing features; re. "possibly confusing" though, here are a few remarks about array creation.
+
+### Syntax for array creation
+
+
+Creating a more-dimensional NumPy array is not that hard -- depending on how you do it. The trick is to use `reshape` to tell NumPy exactly what shape you want. For example, to create an array of all zeros, of dimensions 3 x 4 x 2:
+
+
+```{python}
+np.zeros(24).reshape(4, 3, 2)
+```
+
+But we also want to understand what others might write. And then, you might see things like these:
+
+```{python}
+c1 = np.array([[[0, 0, 0]]])
+c2 = np.array([[[0], [0], [0]]]) 
+c3 = np.array([[[0]], [[0]], [[0]]])
+```
+
+These are all 3-dimensional, and all have three elements, so their shapes must be 1 x 1 x 3, 1 x 3 x 1, and 3 x 1 x 1, in some order. Of course, `shape` is there to tell us:
+
+```{python}
+c1.shape # (1, 1, 3)
+c2.shape # (1, 3, 1)
+c3.shape # (3, 1, 1) 
+```
+
+but we'd like to be able to "parse" internally without executing the code. One way to think about it would be processing the brackets like a state machine, every opening bracket moving one axis to the right and every closing bracket moving back left by one axis. Let us know if you can think of other -- possibly more helpful -- mnemonics!
```

Comment:
```suggestion
but we'd like to be able to "parse" mentally without executing the code. One way to think about it would be processing the brackets like a state machine, every opening bracket moving one axis to the right and every closing bracket moving back left by one axis. Let us know if you can think of other -- possibly more helpful -- mnemonics!
```